File: src/php/sql_injection.php
<?php
/**
 * SQL Injection vulnerabilities for PHP.
 * Used for testing SAST tools.
 */

// Hardcoded credentials
define('DB_USER', 'admin');
define('DB_PASSWORD', 'password123');
define('DB_NAME', 'test');

/**
 * Vulnerable SQL query with string concatenation
 */
function getUserDataVulnerable($userId) {
    $conn = new mysqli('localhost', DB_USER, DB_PASSWORD, DB_NAME);

    // VULNERABLE: Direct string concatenation
    $query = "SELECT * FROM users WHERE id = '$userId'";

    $result = $conn->query($query);

    $users = [];
    while ($row = $result->fetch_assoc()) {
        $users[] = $row;
    }

    $conn->close();
    return $users;
}

/**
 * Multiple parameter injection
 */
function searchUsersVulnerable($name, $email) {
    $conn = new mysqli('localhost', DB_USER, DB_PASSWORD, DB_NAME);

    // VULNERABLE: Multiple parameters concatenated
    $query = "SELECT * FROM users WHERE name = '$name' AND email = '$email'";

    $result = $conn->query($query);

    $users = [];
    while ($row = $result->fetch_assoc()) {
        $users[] = $row;
    }

    $conn->close();
    return $users;
}

/**
 * SQL injection in ORDER BY clause
 */
function getUsersSortedVulnerable($sortColumn) {
    $conn = new mysqli('localhost', DB_USER, DB_PASSWORD, DB_NAME);

    // VULNERABLE: User input in ORDER BY
    $query = "SELECT * FROM users ORDER BY $sortColumn";

    $result = $conn->query($query);

    $users = [];
    while ($row = $result->fetch_assoc()) {
        $users[] = $row;
    }

    $conn->close();
    return $users;
}

/**
 * Command injection vulnerability
 */
function commandInjectionVulnerable($userInput) {
    // VULNERABLE: Command injection
    $command = "ping -c 1 " . $userInput;

    // VULNERABLE: Using shell_exec
    $output = shell_exec($command);

    return $output;
}

/**
 * Path traversal vulnerability
 */
function pathTraversalVulnerable($filename) {
    // VULNERABLE: Path traversal
    $path = "/var/www/uploads/" . $filename;

    $content = file_get_contents($path);

    return $content;
}

/**
 * XSS vulnerability
 */
function xssVulnerable($userInput) {
    // VULNERABLE: No output encoding
    echo "<div>" . $userInput . "</div>";
}

/**
 * File inclusion vulnerability
 */
function fileInclusionVulnerable($page) {
    // VULNERABLE: File inclusion
    include($page . '.php');
}

/**
 * Unserialize vulnerability
 */
function unserializeVulnerable($data) {
    // VULNERABLE: Unserializing user input
    $object = unserialize($data);

    return $object;
}

/**
 * Weak session management
 */
function weakSessionManagement() {
    // VULNERABLE: Weak session ID
    session_id('weak_id');

    // VULNERABLE: Session fixation
    session_start();
}

/**
 * Hardcoded credentials
 */
function checkPasswordVulnerable($input) {
    // VULNERABLE: Hardcoded password
    $password = "Admin@123";

    return $input === $password;
}

/**
 * Weak cryptographic algorithm
 */
function weakCryptoVulnerable($data) {
    // VULNERABLE: MD5 is cryptographically weak
    return md5($data);
}

// Safe examples

/**
 * Safe SQL query using prepared statements
 */
function getUserDataSafe($userId) {
    $conn = new mysqli('localhost', DB_USER, DB_PASSWORD, DB_NAME);

    // SAFE: Using prepared statement
    $stmt = $conn->prepare("SELECT * FROM users WHERE id = ?");
    $stmt->bind_param("s", $userId);
    $stmt->execute();

    $result = $stmt->get_result();

    $users = [];
    while ($row = $result->fetch_assoc()) {
        $users[] = $row;
    }

    $stmt->close();
    $conn->close();

    return $users;
}

/**
 * Safe command execution
 */
function safeCommandExecution($ipAddress) {
    // SAFE: Input validation
    if (preg_match('/^[0-9.]+$/', $ipAddress)) {
        $command = "ping -c 1 " . escapeshellarg($ipAddress);
        $output = shell_exec($command);

        return $output;
    }

    return null;
}

/**
 * Safe path handling
 */
function safePathHandling($filename) {
    // SAFE: Path validation
    $basePath = "/var/www/uploads/";
    $userPath = realpath($basePath . $filename);

    if (strpos($userPath, $basePath) === 0) {
        $content = file_get_contents($userPath);
        return $content;
    }

    throw new Exception("Invalid path");
}

/**
 * Safe XSS prevention
 */
function safeXSS($userInput) {
    // SAFE: HTML escaping
    echo "<div>" . htmlspecialchars($userInput, ENT_QUOTES, 'UTF-8') . "</div>";
}

// Example usage
if (isset($_GET['user_id'])) {
    $users = getUserDataVulnerable($_GET['user_id']);
    print_r($users);
}

?>

File: src/ruby/sql_injection.rb
# SQL Injection vulnerabilities for Ruby.
# Used for testing SAST tools.

require 'sqlite3'
require 'mysql2'

# Hardcoded credentials
DB_USER = 'admin'
DB_PASSWORD = 'password123'
DB_NAME = 'test'

# Vulnerable SQL query with string concatenation
def get_user_data_vulnerable(user_id)
  client = Mysql2::Client.new(
    host: 'localhost',
    username: DB_USER,
    password: DB_PASSWORD,
    database: DB_NAME
  )

  # VULNERABLE: Direct string concatenation
  query = "SELECT * FROM users WHERE id = '#{user_id}'"

  result = client.query(query)

  users = []
  result.each do |row|
    users << row
  end

  users
end

# Multiple parameter injection
def search_users_vulnerable(name, email)
  db = SQLite3::Database.new 'test.db'

  # VULNERABLE: Multiple parameters concatenated
  query = "SELECT * FROM users WHERE name = '#{name}' AND email = '#{email}'"

  result = db.execute(query)

  result
end

# SQL injection in ORDER BY clause
def get_users_sorted_vulnerable(sort_column)
  client = Mysql2::Client.new(
    host: 'localhost',
    username: DB_USER,
    password: DB_PASSWORD,
    database: DB_NAME
  )

  # VULNERABLE: User input in ORDER BY
  query = "SELECT * FROM users ORDER BY #{sort_column}"

  result = client.query(query)

  users = []
  result.each do |row|
    users << row
  end

  users
end

# Command injection vulnerability
def command_injection_vulnerable(user_input)
  # VULNERABLE: Command injection
  command = "ping -c 1 #{user_input}"

  # VULNERABLE: Using backticks or system with user input
  output = `#{command}`

  output
end

# Path traversal vulnerability
def path_traversal_vulnerable(filename)
  # VULNERABLE: Path traversal
  path = "/var/www/uploads/#{filename}"

  content = File.read(path)

  content
end

# XSS vulnerability simulation
def xss_vulnerable(user_input)
  # VULNERABLE: No output encoding
  "<div>#{user_input}</div>"
end

# ERB template injection
def erb_injection_vulnerable(template)
  require 'erb'

  # VULNERABLE: User-controlled template
  erb = ERB.new(template)
  erb.result(binding)
end

# YAML.load vulnerability
def yaml_load_vulnerable(yaml_data)
  require 'yaml'

  # VULNERABLE: YAML.load with user input
  YAML.load(yaml_data)
end

# Marshal.load vulnerability
def marshal_load_vulnerable(data)
  # VULNERABLE: Marshal.load with user input
  Marshal.load(data)
end

# Hardcoded credentials
def check_password_vulnerable(input)
  # VULNERABLE: Hardcoded password
  password = "Admin@123"

  input == password
end

# Weak cryptographic algorithm
def weak_crypto_vulnerable(data)
  require 'digest'

  # VULNERABLE: MD5 is cryptographically weak
  Digest::MD5.hexdigest(data)
end

# Safe examples

# Safe SQL query using parameterized statements
def get_user_data_safe(user_id)
  client = Mysql2::Client.new(
    host: 'localhost',
    username: DB_USER,
    password: DB_PASSWORD,
    database: DB_NAME
  )

  # SAFE: Using parameterized query
  query = "SELECT * FROM users WHERE id = ?"

  result = client.prepare(query).execute(user_id)

  users = []
  result.each do |row|
    users << row
  end

  users
end

# Safe command execution
def safe_command_execution(ip_address)
  # SAFE: Input validation
  if ip_address =~ /^[0-9.]+$/
    # SAFE: Using system with array arguments
    system('ping', '-c', '1', ip_address)
  end
end

# Safe path handling
def safe_path_handling(filename)
  # SAFE: Path validation
  base_path = "/var/www/uploads/"
  user_path = File.expand_path(filename, base_path)

  if user_path.start_with?(File.expand_path(base_path))
    content = File.read(user_path)
    return content
  end

  raise SecurityError, "Invalid path"
end

# Safe XSS prevention
def safe_xss(user_input)
  require 'cgi'

  # SAFE: HTML escaping
  escaped = CGI.escapeHTML(user_input)
  "<div>#{escaped}</div>"
end

# Safe YAML loading
def safe_yaml_load(yaml_data)
  require 'yaml'

  # SAFE: Using safe_load
  YAML.safe_load(yaml_data, permitted_classes: [Symbol, Time])
end

# Example usage
if __FILE__ == $0
  users = get_user_data_vulnerable("1 OR '1'='1'")
  puts users.inspect
end